src/main/java/com/celements/filebase/references/FileReference.java

```diff
@@ -0,0 +1,202 @@
+package com.celements.filebase.references;
+
+import static com.google.common.base.Preconditions.*;
+
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.function.Supplier;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.NotNull;
+import javax.ws.rs.core.UriBuilder;
+
+import org.xwiki.model.reference.DocumentReference;
+
+import com.celements.model.util.ModelUtils;
+import com.google.common.base.Strings;
+import com.google.common.base.Suppliers;
+import com.google.errorprone.annotations.Immutable;
+import com.xpn.xwiki.web.Utils;
+
+@Immutable
+public final class FileReference implements Serializable {
+
+  private static final long serialVersionUID = 1L;
+
+  @NotThreadSafe
+  public static final class Builder {
+
+    private static final String ATTACHMENT_LINK_REGEX = "([\\w\\-]*:)?([\\w\\-]*\\.[\\w\\-]*){1};.*";
+    private static final Supplier<Pattern> ATTACHMENT_LINK_PATTERN = Suppliers
+        .memoize(() -> Pattern.compile(ATTACHMENT_LINK_REGEX));
+    private static final String ON_DISK_LINK_REGEX = "^:[/\\w\\-\\.]*";
+    private static final Supplier<Pattern> ON_DISK_LINK_PATTERN = Suppliers
+        .memoize(() -> Pattern.compile(ON_DISK_LINK_REGEX));

```

Comment:
why do we need a memoize here? The patterns can be intialised statically. Also the regex's can be inlined:
```suggestion
  private static final Pattern ATTACHMENT_LINK_PATTERN = Pattern.compile(
      "([\\w\\-]*:)?([\\w\\-]*\\.[\\w\\-]*){1};.*");
  private static final Pattern ON_DISK_LINK_PATTERN = Pattern.compile("^:[/\\w\\-\\.]*");
```